compiler/rustc_ty_utils/src/layout.rs

```diff
@@ -787,27 +787,24 @@ fn layout_of_uncached<'tcx>(
                 let (start, end) = cx.tcx.layout_scalar_valid_range(def.did());
                 match st.abi {
                     Abi::Scalar(ref mut scalar) | Abi::ScalarPair(ref mut scalar, _) => {
-                        // the asserts ensure that we are not using the
-                        // `#[rustc_layout_scalar_valid_range(n)]`
-                        // attribute to widen the range of anything as that would probably
-                        // result in UB somewhere
-                        // FIXME(eddyb) the asserts are probably not needed,
-                        // as larger validity ranges would result in missed
+                        // Enlarging validity ranges would result in missed
                         // optimizations, *not* wrongly assuming the inner
-                        // value is valid. e.g. unions enlarge validity ranges,
+                        // value is valid. e.g. unions already enlarge validity ranges,
                         // because the values may be uninitialized.
```

Comment:
```suggestion
                        // because the values may be uninitialized.
                        //
                        // Because of that we only check that the start and end
                        // of the range is representable with this scalar type.
```
I think this is what's happening here